feat(web): Add album filter to search

xCJPECKOVERx opened the pull request:
## Description

Adds an album dropdown to the search filter modal, using the recently added album search filters (#18985).
The filter modal is slightly rearranged, with albums and tags both being half-width, taking up the space tags used to.
The ComboBox has been modified to show thumbnails (default off, same layout as usual).
The "Not in any album" filter checkbox is automatically unchecked and disabled if the album filter is set. Clearing albums re-enables the checkbox.

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

1. Create albums, some that share images
2. Search one album, combined with other filters (ie. Person 1, album 1, will display assets with the person in selected album).
3. Searching multiple albums limits the search to assets that are present in both albums (asset in album1 AND album 2).

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

<!-- Images go below this line. -->
![image](https://github.com/user-attachments/assets/44589c81-eb5e-4c3c-89ac-4e2f3f3d6d90)


</details>

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

xCJPECKOVERx committed:
- updated dtos
- added inAlbums to search builder
- only check isNotInAlbum if albumIds is blank/empty

xCJPECKOVERx committed:
- make open-api-typescript
- Allow combobox to have thumbnails
- Add the search-albums-section
- Update search-display-section to disable isNotInAlbum checkbox when albums are selected
- In search filter modal, put albums and tags side by side. Saves space and neither need to be full width
- Add album names filter chips to search page

xCJPECKOVERx committed:
- cleanup combobox thumbnails, use component prop to enable

xCJPECKOVERx committed:
- consider inAlbums as OR

xCJPECKOVERx committed:
Merge branch 'feat/search-filter-album/server' into feat/search-filter-album/web

xCJPECKOVERx committed:
- make open-api-dart

xCJPECKOVERx committed:
- lint & format

xCJPECKOVERx committed:
Merge branch 'feat/search-filter-album/server' into feat/search-filter-album/web

xCJPECKOVERx committed:
- disable search-album-section selectedAlbums chip no-wrap. If a very long album name is selected, it would run into the tags section.

xCJPECKOVERx committed:
- sort the selected albums by albumName descending. Allows for better stacking when albums with very long names are selected.

xCJPECKOVERx committed:
- clean up chip wrapping to no-wrap, overflow ellipsis

xCJPECKOVERx committed:
- remove inAlbums groupBy clause

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/server

xCJPECKOVERx committed:
- merge main open-api

xCJPECKOVERx committed:
- make open-api

xCJPECKOVERx committed:
Merge branch 'feat/search-filter-album/server' into feat/search-filter-album/web

xCJPECKOVERx committed:
- inAlbums filter AND instead of OR

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/server

xCJPECKOVERx committed:
Merge branch 'feat/search-filter-album/server' into feat/search-filter-album/web

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/web

xCJPECKOVERx committed:
- remove if-tags from around search-albums-section
- Add translation to remove filter button tooltip

xCJPECKOVERx committed:
- format

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/web

xCJPECKOVERx commented on web/src/lib/components/shared-components/search-bar/search-albums-section.svelte:
```diff
@@ -0,0 +1,100 @@
+  let { selectedAlbums = $bindable() }: Props = $props();
+
+  let allAlbums: AlbumResponseDto[] = $state([]);
+  let albumMap = $derived(Object.fromEntries(allAlbums.map((album) => [album.id, album])));
```
Really just following how the selection is done in the search-tags-section. I could update both?

xCJPECKOVERx replied:
Looking at more closely, I think the Object.fromEntries is best for this case. It creates a dictionary instead of an array, so the map can be referenced by albumId instead of having to filter/find when handling the selected albums.

xCJPECKOVERx commented on web/src/lib/components/shared-components/search-bar/search-albums-section.svelte:
```diff
@@ -0,0 +1,100 @@
+  let albumMap = $derived(Object.fromEntries(allAlbums.map((album) => [album.id, album])));
+  let selectedOption = $state(undefined);
+
+  let sortedSelectedAlbums = $derived(
+    Array.from(selectedAlbums).sort((a, b) => albumMap[b]?.albumName?.length - albumMap[a]?.albumName?.length),
+  );
```
Arguably a separate PR to add pagination & lazy-loading the combobox? I get the limitation though, fetching 100s of albums potentially when opening the search modal isn't ideal. 
However, doesn't tags technically have this same issue?

mertalev replied:
> Sorting on the client is probably not the best idea. Generally the getAllAlbums endpoint quite sucks (it isn't paginated for instance). I'd like to see a new endpoint that's paginated and takes a sort order. @mertalev @jrasm91 thoughts?

We'd have to make the queries not suck though, and that's easier said than done. It doesn't really make sense to paginate an endpoint that has to do so much work to fetch any data. As soon as you fetch the second page you've spent a lot more time than if you fetched everything from the start.

danieldietzler replied:
Yeah it'd make sense if those were separate PRs. 100s would probably even be fine, but some people have thousands of albums where this becomes infeasible. If you want to see a negative example try the author combobox on the pull requests page lmfao.
I haven't checked but it's possible tags currently suffer from this too. However I don't think that's an excuse to make the same mistake twice ;)

danieldietzler replied:
Ideally the UI lib combobox (https://ui.immich.app/components/multi-select) would support API-supported search, not sure if it currently does.

mertalev replied:
> 100s would probably even be fine, but some people have thousands of albums where this becomes infeasible

As long as the combobox is virtual and it doesn't request *all* their thumbnails at once, fetching thousands of albums is not really an issue and likely still cheaper than doing those DB-level aggregations multiple times.

But if the asset count, start date or end date are not needed, then a paginated endpoint without those aggregations would probably be the most efficient option.

xCJPECKOVERx replied:
> Currently have 4780 tags
FWIW, forcing a thumbnail to load for all of these also has a negligible effect. Though I'm unsure if that's simply because of caching

danieldietzler replied:
> So a good start could be to add a slim/paginated endpoint, that doesn't perform the extra metadata fetch?

I think just in general combobox items (especially if you assume there may be many entries) should be as slim as possible, yes. Every additional that may be needed can be fetched when it's *actually* needed.

mertalev replied:
> I'll add a slim getAllAlbums endpoint (or I could add an option to getAllAlbums that will skip over the metadata fetch, whichever you prefer)

A separate endpoint would be better imo

xCJPECKOVERx committed:
New getAllAlbumsSlim endpoint
- Add new endpoint
- Add slim option to albumService.getAll (default false)
- Use getAllAlbumsSlim in search-albums-section

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/web

xCJPECKOVERx committed:
-open-api
-cleanup

xCJPECKOVERx committed:
fix test

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/web

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/web

xCJPECKOVERx committed:
- correct permission types
- merge

danieldietzler reviewed (commented):
This is probably good enough for the time being, although I'd love to see how badly it performs on large instances with many albums

bo0tzz commented:
Closes #12258 if merged

xCJPECKOVERx committed:
Merge branch 'main' into feat/search-filter-album/web

xCJPECKOVERx committed:
- fix icon
- update openapi-specs

xCJPECKOVERx committed:
open-api
